src/Logic/UsersHelper.php

```diff
@@ -0,0 +1,243 @@
+	 *
+	 * Note that you *have* to provide one of the following fields: 'user_email', 'user_login', 'user_nicename', 'display_name'.
+	 *
+	 * The difference to wp_insert_user is that this method will try to create a user even if the data array is incomplete,
+	 * so email, user login, and user nicename will be generated if not provided. Care is taken to avoid duplicates and user facing
+	 * values that would leak emails or other sensitive information.
+	 *
+	 * @param array $data The data to create the user with. If the 'role' key is present, the user will be assigned that role.
+	 *
+	 * @throws Exception If the user could not be created.
+	 * @throws InvalidArgumentException If the data array is empty or if the 'role' key is in the array and does not contain a valid role. .
+	 */
+	public static function create_or_get_user( array $data ): WP_User {
+		if ( empty( $data ) ) {
+			throw new InvalidArgumentException( 'Data array is empty. Refusing to create user from nothing.' );
+		}
+
+		// Trim all that we can (so strings).
+		$data = array_map( fn( $value ) => is_string( $value ) ? trim( $value ) : $value, $data );
+
+		$vital_fields = [ 'user_email', 'user_login', 'user_nicename', 'display_name' ];
+		// We need at least one of the above fields to be present, so check that filtering on not empty does not produce an empty array.
+		if ( empty( array_filter( $vital_fields, fn( $field ) => ! empty( $data[ $field ] ) ) ) ) {
+			// phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+			throw new InvalidArgumentException( sprintf( 'Data array is missing one or more of the vital fields: %s.', implode( ', ', $vital_fields ) ) );
+		}
+
+		if ( ! empty( $data['role'] ) && null === get_role( $data['role'] ) ) {
+			// phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+			throw new InvalidArgumentException( sprintf( 'Role "%s"does not exist.', $data['role'] ) );
+		}
+
+		$wp_user = self::get_user( $data );
+		if ( $wp_user ) { // Great – we already have the user!
+			if ( ! empty( $data['role'] ) ) {
+				// If the role was passed in the data array – add it before returning.
+				$wp_user->add_role( $data['role'] );
+			}
+
+			return $wp_user;
+		}
+
+		// For creating the user, these fields in particular need to be processed, so get the variables here.
+		$user_email    = $data['user_email'] ?? '';
+		$user_nicename = $data['user_nicename'] ?? '';
+		$user_login    = $data['user_login'] ?? '';
+
+
+		// If we don't have an email, we'll create an ugly unusable one so that we can create the user.
+		if ( empty( $user_email ) ) {
+			$email_domain = apply_filters( 'nmt_user_email_default_domain', 'example.com' );
+			$user_email   = self::get_short_sha_from_array( $data ) . '@' . $email_domain;
+		}
+
+		// We insist on setting a nicename to avoid WP setting it to the email sans @.
+		if ( empty( $user_nicename ) ) {
+			$user_nicename = ( $data['first_name'] ?? '' ) . ' ' . ( $data['last_name'] ?? '' );
```

Comment:
Can we change this to like:

```php
$user_nicename = implode( ' ', array_filter( [ $data['first_name'], $data['last_name'] ] ) );
```

Just in case we have a last name without the first name. This way, we won't end up with a space at the beginning of the name.

Reply:
Good call. I'm trimming it now 👍